### PR TITLE
platform(secrets): Support custom detectors from the platform

### PR DESCRIPTION
**Pull request by lirshindalman**
Move custom regex detector to checkov/secrets
## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my feature, policy, or fix is effective and works
- [x] New and existing tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: checkov/secrets/plugins/custom_regex_detector.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+
+def get_detectors_from_cache(customer_name: str | None) -> list[dict[str, Any]]:
+    if customer_name:
+        cache = get_customer_cache()
+        return cache.get(customer_name, [])
+    return []
+
+
+def get_detectors_from_local_file() -> list[dict[str, Any]]:
```
Comment (maxamel) [inline checkov/secrets/plugins/custom_regex_detector.py:33]: Is this method called from anywhere?

File: checkov/secrets/plugins/custom_regex_detector.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+def load_detectors() -> list[dict[str, Any]]:
+    customer_name = os.getenv('CUSTOMER_NAME')
+    detectors = get_detectors_from_cache(customer_name)
+    if not detectors:
+        try:
+            customer_run_config_response = bc_integration.customer_run_config_response
+            policies_list:  List[dict[str, Any]] | dict[str, Any] = customer_run_config_response['secretsPolicies'] if \
+                customer_run_config_response['secretsPolicies'] else []
+        except Exception as e:
+            return []
```
Comment (maxamel) [inline checkov/secrets/plugins/custom_regex_detector.py:48]: Perhaps add a log here to know what went wrong?

File: checkov/secrets/plugins/custom_regex_detector.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+from detect_secrets.core.potential_secret import PotentialSecret
+from detect_secrets.plugins.base import RegexBasedDetector
+import re
+
+from checkov.common.bridgecrew.platform_integration import bc_integration
+from detect_secrets.util.code_snippet import CodeSnippet
+from detect_secrets.util.inject import call_function_with_arguments
+
+
+DETECTORS_BY_CUSTOMER_CACHE: dict[str, list[dict[str, Any]]] = {}
```
Comment (maxamel) [inline checkov/secrets/plugins/custom_regex_detector.py:19]: Do we want to handle cache invalidation in some way?
Meaning, add timestamp and invalidate too old entries?

File: checkov/secrets/plugins/custom_regex_detector.py (hunk 1 above)
Comment (lirshindalman) [reply checkov/secrets/plugins/custom_regex_detector.py:33]: not anymore, you're right, I will delete it

File: checkov/secrets/runner.py [hunk 4]
```diff
@@ -95,18 +96,16 @@ def run(
...
-            if exists(detector_path):
-                logging.info(f"Custom detector found at {detector_path}. Loading...")
-                plugins_used.append({
-                    'name': 'CustomRegexDetector',
-                    'path': f'file://{detector_path}'
-                })
-            else:
-                logging.info(f"Custom detector not found at path {detector_path}. Skipping...")
+
+        detector_path = f"{current_dir}/plugins/custom_regex_detector.py"
```
Comment (maxamel) [inline checkov/secrets/runner.py:100]: Is there anything preventing us from just loading the file? The "exists" check was used when the CustomRegexDetector was injected externally. Now that the custom detector is part of checkov I can't think of a scenario where this file would not exist, no ?

Comment (lirshindalman) [reply checkov/secrets/runner.py:100]: yes you right

**Review by maxamel — APPROVED** [review]

File: checkov/secrets/plugins/custom_regex_detector.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+import yaml
+from detect_secrets.constants import VerifiedResult
+from detect_secrets.core.potential_secret import PotentialSecret
+from detect_secrets.plugins.base import RegexBasedDetector
+from detect_secrets.util.code_snippet import CodeSnippet
+from detect_secrets.util.inject import call_function_with_arguments
+import re
+
+from checkov.common.bridgecrew.platform_integration import bc_integration
```
Comment (gruebel) [inline checkov/secrets/plugins/custom_regex_detector.py:15]:
```suggestion
import logging
from abc import ABC
from typing import Set, Any, Generator, Pattern, Optional, Dict, Tuple, List, TYPE_CHECKING

import yaml
from detect_secrets.constants import VerifiedResult
from detect_secrets.core.potential_secret import PotentialSecret
from detect_secrets.plugins.base import RegexBasedDetector
from detect_secrets.util.inject import call_function_with_arguments
import re

from checkov.common.bridgecrew.platform_integration import bc_integration

if TYPE_CHECKING:
    from detect_secrets.util.code_snippet import CodeSnippet
```

File: checkov/secrets/plugins/custom_regex_detector.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    return custom_detectors
+
+
+class CustomRegexDetector(RegexBasedDetector, ABC):
+
+    denylist: Set[Pattern[str]] = set()
+
+    def __init__(self) -> None:
+        self.regex_to_metadata: dict[str, dict[str, Any]] = dict()
+        self.denylist = set()
```
Comment (gruebel) [inline checkov/secrets/plugins/custom_regex_detector.py:69]:
```suggestion
        self.denylist: set[Pattern[str]] = set()
```

File: checkov/secrets/runner.py [hunk 7]
```diff
@@ -95,18 +95,13 @@ def run(
...
+        logging.info(f"Custom detector found at {detector_path}. Loading...")
+        plugins_used.append({
+            'name': 'CustomRegexDetector',
+            'path': f'file://{detector_path}'
```
Comment (lirshindalman) [inline checkov/secrets/runner.py:103]: @nimrodkor

File: checkov/secrets/runner.py [hunk 8]
```diff
@@ -95,18 +95,15 @@ def run(
...
-                    'name': 'CustomRegexDetector',
-                    'path': f'file://{detector_path}'
-                })
-            else:
-                logging.info(f"Custom detector not found at path {detector_path}. Skipping...")
+
+        detector_path = f"{current_dir}/plugins/custom_regex_detector.py"
+        logging.info(f"Custom detector found at {detector_path}. Loading...")
+        enable_secret_scan_all_files = runner_filter.enable_secret_scan_all_files
+        if enable_secret_scan_all_files:
```
Comment (lirshindalman) [inline checkov/secrets/runner.py:102]: @maxamel  @nimrodkor

**Review by nimrodkor — APPROVED** [review]

**Pull request merged**